stdlib: Improve Never documentation

bradleymackey opened the pull request:
<!-- What's in this pull request? -->
Reopens #38004, with suggested changes applied.

bradleymackey committed:
Improve Never documentation

bradleymackey commented:
@AnthonyLatsis Hi, could you ask the library folks to look at this? Thanks!

harlanhaskins commented on stdlib/public/core/Policy.swift:
```diff
@@ -15,16 +15,42 @@
-/// method that unconditionally throws an error, traps, or otherwise does
-/// not terminate.
+/// Nonreturning functions and methods either cause an irrecoverable error
+/// or begin a sequence of work that continues indefinitely. This means that
```
I think this sentence got

bradleymackey replied:
Yikes, that could have been

lorentey reviewed (approved):
The technical parts looks good to me! (I'm not an expert on Swift's documentation style, though. E.g., there may be house rules around the use of em dashes or italics.) Cc @amartini51

lorentey commented:
@swift-ci smoke test

amartini51 commented:
Overall, this looks like an improvement because it adds information that was missing before.  However, there are several style issues.  Do you have a preference between me rewriting and committing changes versus suggesting changes in PR comments?  For example, the abstract uses italics and defines new terms, and we usually avoid parenthetical asides.

There's also a question of positioning.  Which information is more important to a developer reading this documentation: The type theory around `Never` being an uninhabited type, or the use cases like `Never` being the type you use as the return value for functions that don't return?

amartini51 committed:
Revise docs for Never for reference style.

- Reduce the emphasis on the type theory that Never is an uninhabited
  type, focusing more on its meaning and usage in code.
- Move the definition of uninhabited out of the abstract.  Define
  "nonreturning" more explicitly.
- Expand the favoriteNumber example's code comment into a brief
  paragraph to walk through the code listing.
- Avoid italics in the abstract, future tense, and parenthetical asides.
- Use contractions.

bradleymackey commented on stdlib/public/core/Policy.swift:
```diff
@@ -15,16 +15,49 @@
-/// method that unconditionally throws an error, traps, or otherwise does
-/// not terminate.
+/// Use `Never` as the return type of a function
+/// that doesn't ever return normally --- for example,
```
I think eliding "ever" reads slightly better to me here, but I acknowledge the benefit of using an antonym of "never" to clarify the semantics here, especially to non-native English speakers.

```suggestion
/// that doesn't return normally --- for example,
```

amartini51 replied:
You're correct, I got to this wording like you said, by describing the behavior without repeating the word "never".  But your revision looks good.

bradleymackey commented on stdlib/public/core/Policy.swift:
```diff
@@ -15,16 +15,49 @@
+///     }
+///
+/// In the code above,
+/// the result type has a failure type of `Never`,
```
Repetition of "type" doesn't read so well to me, so perhaps better to reference the name of the variable directly rather than referencing it's type? 

```suggestion
/// `favoriteNumber` has a failure type of `Never`,
```

amartini51 replied:
I think this works.  Either way, we're omitting one step in a three-element chain — that is, favoriteNumber's result's failure type is `Never`.  Since the `Result` is closer to the `Never`, that what I used, but I agree with your point about not repeating "type" too many times.  Your rewrite might be a bit easier to follow too, since it starts at the beginning of the three-element chain.

bradleymackey reviewed (commented):
Thanks @amartini51 for looking into this. I think your suggestions are a substantial improvement. I'd consider the non-returning nature of `Never` to be the most important feature of the documentation as that's where new developers are more likely to encounter it for the first time (e.g. `fatalError`). Delaying the introduction of the term "uninhabited" to below the non-returning examples seems like a sensible choice to me!

amartini51 committed:
Avoid repeating "type".

Co-authored-by: Bradley Mackey <bradley@mcky.dev>

amartini51 committed:
Tighten phrasing.

Co-authored-by: Bradley Mackey <bradley@mcky.dev>

amartini51 commented:
@swift-ci Please smoke test.

lorentey commented:
@swift-ci smoke test macOS platform

bradleymackey committed:
Fix signature of fatalError

bradleymackey committed:
Format to 80w

amartini51 commented:
@bradleymackey I'd prefer you don't hard-wrap this content back to 80 columns (abf877ada2dde1fa402bf584ac6c2ef60edd0cd4), although if you feel very strongly on this point I'll concede it.  Hard wrapping documentation means each time a change is made, the entire paragraph must be re-wrapped.  That produces noisy diffs, making docs PRs harder to review, and reduces the Git history available in `git blame`.

To prevent this, several areas in documentation use semantic line breaks — that is, each line contains a meaningful phrase and doesn't need to be rewrapped at each revision.  That approach lets you get more granular Git history and produces much smaller diffs when making changes, which makes documentation PRs easier to review.

I've written some a longer comments about semantic line breaks in the style guide for "The Swift Programming Language" [here](https://github.com/apple/swift-book/blob/main/Style.md#semantic-line-breaks).

bradleymackey commented:
@amartini51 That makes a lot of sense, I'll revert the change. Thanks for explaining!

bradleymackey committed:
Revert "Format to 80w"

This reverts commit abf877ada2dde1fa402bf584ac6c2ef60edd0cd4.

amartini51 commented:
@swift-ci Please smoke test macOS platform

amartini51 commented:
@swift-ci Please smoke test

amartini51 commented:
@swift-ci Please smoke test macOS
